### PR TITLE
added basic controller for librarian service.

### PR DESCRIPTION
**Pull request by sean-darsie**
Made endpoints for each of the methods in the librarian service class in the librarian controller.

### CODE REVIEW

**Review by trevorhuis-smoothstack — APPROVED** [review]
I see you used the pathVariables instead of query parameters. I'm not sure which one is the correct style if you have multiple variables but we should decide on a strategy. I will build with pathVariables for now

**Review by justin-obrien-smoothstack — CHANGES_REQUESTED** [review]
In the event of an SQLException, it seems to me the best status code is 500 (Internal Server Error). If an SQLException was thrown it basically means one of our database access methods failed, and the client's request may have been fine.

I'm pretty sure each controller method should return a status code to the client; updateBranch and updateCopies currently return void.

I think methods should generally return 404 if the client requests something that doesn't exist in the database (of course, exceptions exist, like if a librarian wants to add copies of a book that their branch doesn't have any copies of yet).

It seems that it's best to have the service classes start throwing exceptions instead of catching them, so the service controllers can catch the exceptions and return the appropriate status codes accordingly. Currently the service classes print to the terminal in response to exceptions, but that won't do much good within our current paradigm.

File: src/main/java/com/ss/training/lms/controller/LibrarianController.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+        try {
+            bookCopies = librarianService.getAnEntryOfBookCopies(branch, book);
+            if (bookCopies == null)
+            {
+                status = HttpStatus.NOT_FOUND;
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+            return new ResponseEntity<BookCopies>(bookCopies , HttpStatus.BAD_REQUEST);
```
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/lms/controller/LibrarianController.java:93]: It looks like you missed this one when changing the Bad Requests to Internal Server Errors.

File: src/main/java/com/ss/training/lms/controller/LibrarianController.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    {
+        BookCopies bookCopies = null;
+        HttpStatus status = HttpStatus.OK;
+
+        try {
+            bookCopies = librarianService.getAnEntryOfBookCopies(branch, book);
+            if (bookCopies == null)
+            {
+                status = HttpStatus.NOT_FOUND;
+            }
```
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/lms/controller/LibrarianController.java:90]: Here's something tricky: What if they want to add copies of a book that does exist in the database, but the book copies table has no entry for that book at that branch? Ideally we'd want to distinguish that from the case where the request specifies a book or branch that doesn't exist in the database. I'm thinking if the copies object is null at this point, we should check to see if the book and branch both exist, and then respond accordingly.

**Review by justin-obrien-smoothstack — CHANGES_REQUESTED** [review]
If I'm not mistaken, the service class methods still have some error-handling built in. We want all those errors to propagate up to the controller so it can handle them.

I don't think the printStackTrace calls in the catch blocks will do anything.

I think one thing to consider, but not strictly necessary, is returning No Content instead of Not Found when there are no books at their branch or their book search returns no results. In both cases they're asking us what books that meet a certain criteria, and we're answering their question by telling them there are no such books -- they asked what the content is, and the answer is that there is no content. However I do think Not Found still makes the most sense whent where are no branches, because presumably that method gets run for the purpose of allowing the librarian to choose their branch, which is necessary before they can do anything else, so there certainly should be available branches when that happens.

File: src/main/java/com/ss/training/lms/controller/LibrarianController.java (hunk 2 above)
Comment (sean-darsie) [reply src/main/java/com/ss/training/lms/controller/LibrarianController.java:90]: The service class actually already handles this. It checks if a record of that book exists at the branch. if so then it does an update. If not it does an add

**Pull request merged**